wallet/src/de/schildbach/wallet/ui/send/PaymentProtocolViewModel.kt

```diff
@@ -127,14 +126,22 @@ class PaymentProtocolViewModel(application: Application) : SendCoinsBaseViewMode
         signAndSendPayment(finalPaymentIntent!!, baseSendRequest!!.ensureMinRequiredFee)
     }
 
-    fun directPay(transaction: Transaction) {
+    override fun signAndSendPayment(sendRequest: SendRequest, txAlreadyCompleted: Boolean) {
```

Comment:
I am not sure what `txAlreadyCompleted` means.  On line 126 above, `txAlreadyCompleted` is set to `ensureMinRequiredFee` which I thought was usually set to true.

Reply:
It is pretty tricky ;)
Before sending the transaction to server it has to be completed. Previously it was completed by the `sendCoinsOffline` method since we were broadcasting the transaction before sending to server. Now as we first want to send it to the server it has to be completed explicitly (and it is in https://github.com/dashevo/dash-wallet/pull/502/files/af444fa05fec1e0f6e56b242a7f5394f0707b94a#diff-6c4b3d3eb5b78806c89f5e5ffda4e1a1R131). On the other hand it cannot be completed again when we will be broadcasting the transaction. That is why I added this flag, but you probably already know that.
The tricky part is related to inheritance of payment related ViewModels. In the mentioned line 126, other version of this method is being called, it gets PaymentIntent as a first parameter, not the SendRequest.
The only place where the `txAlreadyCompleted` flag is true is here:
```
fun commitAndBroadcast(sendRequest: SendRequest) {
        super.signAndSendPayment(sendRequest, true)
}
```
https://github.com/dashevo/dash-wallet/pull/502/files/af444fa05fec1e0f6e56b242a7f5394f0707b94a#diff-6c4b3d3eb5b78806c89f5e5ffda4e1a1R168